### PR TITLE
A-658 Merge PartialCodec and Codec into a single 'Codec' entity

### PR DESCRIPTION
**Pull request by marcesquerra**

### CODE REVIEW

File: src/io/iohk/decco/BufferInstantiatior.scala [hunk 1]
```diff
@@ -0,0 +1,62 @@
+package io.iohk.decco
```
Comment (jtownson) [inline src/io/iohk/decco/BufferInstantiatior.scala]: Typo in filename (too many 'i's).

**Review by jtownson — CHANGES_REQUESTED** [review]

File: src/io/iohk/decco/BufferInstantiatior.scala (hunk 1 above)
Comment (marcesquerra) [reply src/io/iohk/decco/BufferInstantiatior.scala]: Good catch, thanks!

Fixed :)

File: src/io/iohk/decco/Codec.scala [hunk 2]
```diff
@@ -2,106 +2,99 @@ package io.iohk.decco
...
-      } else {
-        Left(BodyTooShort(source.remaining - nextIndex, sizeField))
+  final def zip[U](that: Codec[U]): Codec[(T, U)] = new Codec[(T, U)] {
+
+    def size(tu: (T, U)): Int = tu match {
+      case (t, u) => self.size(t) + that.size(u)
+    }
+
+    def decodeImpl(start: Int, source: ByteBuffer): Either[Failure, DecodeResult[(T, U)]] = {
+      self.decodeImpl(start, source) match {
```
Comment (AlexITC) [inline src/io/iohk/decco/Codec.scala:36]: Looks like we can use for-comprehension.

File: src/io/iohk/decco/Codec.scala [hunk 3]
```diff
@@ -2,106 +2,99 @@ package io.iohk.decco
...
-  def directCodec[T](implicit pc: PartialCodec[T], typeCode: TypeCode[T]): Codec[T] = new Codec[T](pc, typeCode) {
-    override def newBuffer(capacity: Int): ByteBuffer = ByteBuffer.allocateDirect(capacity)
+  final def mapOpt[U](t2u: T => Option[U], u2t: U => T): Codec[U] = new Codec[U] {
+    def size(u: U): Int = self.size(u2t(u))
+    def decodeImpl(start: Int, source: ByteBuffer): Either[Failure, DecodeResult[U]] =
+      self.decodeImpl(start, source) match {
+        case Left(Failure) =>
+          Left(Failure)
+        case Right(DecodeResult(t, nextIndex)) =>
+          t2u(t) match {
```
Comment (AlexITC) [inline src/io/iohk/decco/Codec.scala:78]: an alternative: `t2u(t).map { u => Right(DecodeResult(u, nextIndex)) }.toRight(Failure)`

File: src/io/iohk/decco/Codec.scala [hunk 4]
```diff
@@ -2,106 +2,99 @@ package io.iohk.decco
...
+  type Failure = Failure.type
 
-  // 'Rehydrating' function.
-  // From a buffer of unknown type, read the typeCode and apply the
-  // data to a function that maps to a decoder with the corresponding typeCode.
-  def decodeFrame(decoderWrappers: Map[String, (Int, ByteBuffer) => Unit], start: Int, source: ByteBuffer): Unit = {
-    val hashWrappers: Map[MD5, (Int, ByteBuffer) => Unit] = decoderWrappers.map {
-      case (typeCode, decoderWrapper) => (MD5(typeCode), decoderWrapper)
-    }
+  def apply[T](implicit ev: Codec[T]): Codec[T] = ev
```
Comment (AlexITC) [inline src/io/iohk/decco/Codec.scala:98]: nice trick

**Review by AlexITC — APPROVED** [review]
Now we just need another refactor that uses a backwards-compatibility protocol, like protobuf.

**Pull request merged**